source/Calamari/LaunchTools/NodeExecutor.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.IO;
+using Calamari.Common.Commands;
+using Calamari.Common.Features.Processes;
+using Calamari.Common.Plumbing;
+using Calamari.Common.Plumbing.Commands;
+using Calamari.Common.Plumbing.Extensions;
+using Calamari.Common.Plumbing.FileSystem;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.Variables;
+
+namespace Calamari.LaunchTools
+{
+    [LaunchTool(LaunchTools.Node)]
+    public class NodeExecutor : LaunchTool<NodeInstructions>
+    {
+        readonly CommonOptions options;
+        readonly IVariables variables;
+        readonly ICommandLineRunner commandLineRunner;
+        readonly ILog log;
+
+        public NodeExecutor(CommonOptions options, IVariables variables, ICommandLineRunner commandLineRunner, ILog log)
+        {
+            this.options = options;
+            this.variables = variables;
+            this.commandLineRunner = commandLineRunner;
+            this.log = log;
+        }
+
+        protected override int ExecuteInternal(NodeInstructions instructions, params string[] args)
+        {
+            var pathToNode = variables.Get(instructions.NodePathVariable);
+            var pathToStepPackage = variables.Get(instructions.TargetPathVariable);
+            var pathToBootstrapper = variables.Get(instructions.BootstrapperPathVariable);
+            var runningDeployment = new RunningDeployment(variables);
+
+            var copyVariables = new CalamariVariables();
+            variables.GetNames().ForEach(name => copyVariables.Set(name, variables.Get(name)));
+
+            using (var variableFile = new TemporaryFile(Path.GetTempFileName()))
+            {
+                var variablesAsJson = copyVariables.SaveAsString();
```

Comment:
We have to serialize the variables again because any previous steps could have created output variables and we want those available in subsequent steps that are run out of proc